File: src/taller/taller3/RechargeableSubwayCard.java
package taller.taller3;

public class RechargeableSubwayCard extends SubwayCard {
    private double balance;

    RechargeableSubwayCard(SubwayCentral sc) {
        super(sc);
    }

    public void recharge(double amount) {
        balance += amount;
    }

    protected double rideCost() {
        SubwayCentral sc = getSubwayCentral();
        return sc.getRideCost();
    }

    @Override
    protected boolean canRide() {
        return balance >= rideCost();
    }

    @Override
    public void ride() {
        if(canRide()) {
            SubwayCentral sc = getSubwayCentral();
            balance -= rideCost();
        }
        else {
            System.out.println("No se puede viajar");
        }
    }
}

File: POO-Ruby/guias/guia8/ej6/rechargeable_subway_card.rb
# frozen_string_literal: true
require_relative 'subway_card'
require_relative 'subway_central'

class RechargeableSubwayCard < SubwayCard
  def initialize(subway_central)
    @subway_central = subway_central
    @balance = 0
  end

  def recharge(amount)
    raise 'Invalid amount' unless amount.is_a?(Numeric) && amount.positive?
    @balance += amount
  end

  private def ride_cost
    @subway_central.ride_cost
  end

  private def ride?
    @balance >= self.ride_cost
  end

  def ride
    raise 'not enough funds' unless ride?
    @balance -= ride_cost
  end

  def to_s
    "Rechargeable #{super} [Balance: $#{@balance}]"
  end

end
